component/http/cache/cache.go

```diff
@@ -0,0 +1,349 @@
+type cacheControl struct {
+	noCache         bool
+	forceCache      bool
+	warning         string
+	validators      []validator
+	expiryValidator validator
+}
+
+// executor is the function returning a cache Response object from the underlying implementation
+type executor func(now int64, key string) *cachedResponse
+
+// handler wraps the an execution logic with a cache layer
+// exec is the processor func that the cache will wrap
+// rc is the route cache implementation to be used
+func handler(exec executor, rc *RouteCache) func(request *handlerRequest) (response *handlerResponse, e error) {
+
+	return func(request *handlerRequest) (response *handlerResponse, e error) {
+
+		now := NowSeconds()
+
+		key := request.getKey()
+
+		var rsp *cachedResponse
+
+		if hasNoAgeConfig(rc.age.min, rc.age.max) {
+			rsp = exec(now, key)
+			return &rsp.Response, rsp.Err
+		}
+
+		cfg := extractRequestHeaders(request.header, rc.age.min, rc.age.max-rc.age.min)
+		if cfg.expiryValidator == nil {
+			cfg.expiryValidator = expiryCheck
+		}
+
+		rsp = getResponse(cfg, request.path, key, now, rc, exec)
+		e = rsp.Err
+
+		if e == nil {
+			response = &rsp.Response
+			addResponseHeaders(now, response.Header, rsp, rc.age.max)
+			if !rsp.FromCache && !cfg.noCache {
+				saveToCache(request.path, key, rsp, rc.cache, time.Duration(rc.age.max)*time.Second)
+			}
+		}
+
+		return
+	}
+}
+
+// getResponse will get the appropriate Response either using the cache or the executor,
+// depending on the
+func getResponse(cfg *cacheControl, path, key string, now int64, rc *RouteCache, exec executor) *cachedResponse {
+
+	if cfg.noCache {
+		return exec(now, key)
+	}
+
+	rsp := getFromCache(key, rc)
+	if rsp == nil {
+		metrics.miss(path)
+		response := exec(now, key)
+		return response
+	}
+	if rsp.Err != nil {
+		log.Errorf("error during cache interaction: %v", rsp.Err)
+		metrics.err(path)
+		return exec(now, key)
+	}
+	// if the object has expired
+	if isValid, cx := isValid(now-rsp.LastValid, rc.age.max, append(cfg.validators, cfg.expiryValidator)...); !isValid {
+		tmpRsp := exec(now, key)
+		// if we could not retrieve a fresh Response,
+		// serve the last cached value, with a Warning Header
+		if cfg.forceCache || tmpRsp.Err != nil {
+			rsp.Warning = "last-valid"
+			metrics.hit(path)
+		} else {
+			rsp = tmpRsp
+			metrics.evict(path, cx, now-rsp.LastValid)
+		}
+	} else {
+		// add any Warning generated while parsing the headers
+		rsp.Warning = cfg.warning
+		metrics.hit(path)
+	}
+
+	return rsp
+}
+
+func isValid(age, maxAge int64, validators ...validator) (bool, validationContext) {
+	if len(validators) == 0 {
+		return false, 0
+	}
+	for _, validator := range validators {
+		if isValid, cx := validator(age, maxAge); !isValid {
+			return false, cx
+		}
+	}
+	return true, 0
+}
+
+// getFromCache is the implementation that will provide a cachedResponse instance from the cache,
+// if it exists
+func getFromCache(key string, rc *RouteCache) *cachedResponse {
```

Comment:
since we are in the cache package could we name if just:
```suggestion
func get(key string, rc *RouteCache) *cachedResponse {
```